Add "recently accessed" to Kibana Home

nreese opened the pull request:
fixes https://github.com/elastic/kibana/issues/16280

<img width="1419" alt="screen shot 2018-02-13 at 10 39 47 am" src="https://user-images.githubusercontent.com/373691/36164667-450b4598-10aa-11e8-91f9-605e786557e0.png">




Had to bump `enzyme-to-json` to `3.3.0` for [React.Fragment enzyme support](https://github.com/adriantoine/enzyme-to-json/issues/89)

cc @alexfrancoeur @elastic/kibana-design

nreese commented:
@formgeist The mockups had the accessed list only showing a couple of values that fit on a single line. What is the best way to calculate what will fit? And to know when the rest need to be displayed in a drop down?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15512/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15516/)

formgeist commented:
@nreese I believe the idea was to have max. 5 before moving the rest into a dropdown. I suppose the truncation you have now doesn't work if each link has percentage width, so it'll be fluid if the screensize is smaller. Probably should have a breakpoint where the items just line up in a list instead, still max. 5 before a dropdown. 

Noticed that when I saved a query it created a "New saved search" item as well, but it seems to be discarded when I try to load it. The proper saved search with my long name works.

![screenshot 2018-02-07 11 41 57](https://user-images.githubusercontent.com/4104278/35912287-7125c9e4-0bfc-11e8-83a9-c8f6d80ab6ab.png)

@snide Do you have any tricks to calculate the sizing responsively like described? ^

cchaos commented:
You should be able to use Flexbox and if you only give the last item (the +5 dropdown) a set width, the others should fill in the rest equally with `flex: 1`. Just ensure the link doesn't line-wrap to allow for truncation.

nreese commented:
@cchaos So I should not be truncating the labels, but instead letting CSS do it?

cchaos commented:
That would be the most flexible option.

nreese commented:
@cchaos Thanks for the help. The recently accessed section now exists on a single line and CSS is use to shorten the link text.

@formgeist Thanks for finding the "New Saved Search" getting added to the history. I have added logic to avoid pushing searches without ids onto the history.

alexfrancoeur commented:
Pretty awesome @nreese!

I'm wondering what the best approach to handling < 5 saved objects are. Right now, it seems that we're consuming the entire space. Would it be better to align left? Looking for design input here. @cchaos @formgeist 

<img width="1220" alt="screen shot 2018-02-07 at 3 56 21 pm" src="https://user-images.githubusercontent.com/25181823/35941231-9534deb0-0c1f-11e8-85e5-dc554c2a768b.png">

> 5 saved objects look good! Are we truncating the name here as well if it is greater than a certain amount of characters?

<img width="1228" alt="screen shot 2018-02-07 at 3 56 46 pm" src="https://user-images.githubusercontent.com/25181823/35941335-e05dd8ce-0c1f-11e8-9e99-0c000e5ff8a2.png">

A few other quick questions.

What did we decide upon for "separating" these links. At first it was `>`, we discussed potential icons. Right now it's just spacing, is that where we landed? @cchaos @formgeist 

@nreese I imagine that after we introduce this feature, users are going to wonder why they're recently visited ML jobs or APM applications aren't in the list. Thoughts on the scope of work there? Are we building in a way that accommodate these plugins as well?

cchaos commented:
Yeah, it would be best to left align, though that gets harder to manage when you're trying to truncate... Somewhat of a complex grid layout here. I don't know the final solution for that.

I believe @formgeist was coming up with a solution for separating items that included dots and/or icons?

nreese commented:
@alexfrancoeur Plugin authors will just have to do the following to add items to the history

```
import { recentlyAccessed } from 'ui/persisted_log';

recentlyAccessed.add('item link', 'item label');
```

alexfrancoeur commented:
Thanks @nreese, I'll reach out to a few different teams.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15547/)

formgeist commented:
Been playing around with some layout changes to clean up everything. If we're putting the each of the Add data panels inside the big 100% panel, we should remove the borders. I think we have a general consensus around not doing panel in panel. I've also just adjusted how we display the "Data already in Elasticsearch? Set up index patterns" option so it flows a little better.

Beyond the separator between each item in the Recently Accessed panel, we discussed using the `dot` icon with its colour subdued, much like how it's presented in the following screenshot. In my screen the items flow nicely, since they're not a fixed width, but rather size whatever the text is. I understand this might be problematic when truncating with CSS like we do, so maybe we can agree a max-width that looks good, and put the `dot` between each item? 

![image](https://user-images.githubusercontent.com/4104278/35974068-f1f638a2-0cd7-11e8-853e-556e38a3937a.png)

cchaos commented:
@nreese [This is a quick and dirty solution for inline truncation](https://codepen.io/clhorn99/pen/KQaMQj). You'd just need to figure out a way to update the more dropdown to say and show whatever is not visible.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15606/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15614/)

nreese commented:
@alexfrancoeur Try out the latest. The CSS has changed to fix the justification problem. The flex items no longer grow to fill the space and the `min-width` is getting dynamically set to avoid shrinking really small names to much.

<img width="1200" alt="screen shot 2018-02-08 at 1 21 40 pm" src="https://user-images.githubusercontent.com/373691/35996458-d6b8437c-0cd3-11e8-8040-e7cccdcc6fea.png">

<img width="1193" alt="screen shot 2018-02-08 at 1 22 06 pm" src="https://user-images.githubusercontent.com/373691/35996490-e3f1e818-0cd3-11e8-9c41-32c806ebd870.png">

<img width="1195" alt="screen shot 2018-02-08 at 1 22 40 pm" src="https://user-images.githubusercontent.com/373691/35996505-ea28cb98-0cd3-11e8-9b31-c92a370c2cf5.png">

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15625/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15628/)

cjcenizal commented:
I encountered an odd bug. Steps to repro:

1. Open an existing visualization.
2. Save it with a new name.
3. Go to Home. You won't see the new visualization in the list of "Recently accessed".

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.js:
```diff
@@ -0,0 +1,188 @@
+    // Dynamically setting the min width based on label lengh meets both of these goals.
+    const EM_RATIO = 0.6; // 'em' ratio that avoids too much horizontal white space and too much truncation
+    const minWidth = (label.length < 8 ? label.length : 8) * EM_RATIO;
+    const style = {
```
Can we create a custom class for these styles instead of using an inline style? This will make it a little easier for engineers and designers to find these styles since they'll live in an expected location (e.g. LESS file or SCSS file once we support it).

Can we also add a `max-width: 300px` or something around that size? That will help avoid these kinds of unbalanced layouts:

![image](https://user-images.githubusercontent.com/1238659/36004707-e39674dc-0ce8-11e8-96b8-5672f3d81c72.png)

cjcenizal replied:
Also, maybe we should add a tooltip to show the full name of each item when you hover?

nreese replied:
How would you dynamically set a min-width via less? A variable min-width really helps avoid lots of extra white space around really short names.

cjcenizal replied:
Ah, shoot, I was skimming and overlooked that requirement. Sorry. Yeah that's a tough one... I think your solution is probably the best way to go about it.

cchaos replied:
@cjcenizal As for the tooltip idea, anything that could ever get truncated needs to have it's text in a title attribute. That way, on hover you will always get the full text.

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.js:
```diff
@@ -0,0 +1,188 @@
+      links.push(this.renderLongLink(
+        this.props.recentlyAccessed[i].link,
+        this.props.recentlyAccessed[i].label,
+        includeSeperator));
```
Hate to be that guy, but this should be `includeSeparator` (sepArate)

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.js:
```diff
@@ -0,0 +1,188 @@
+  renderLongLink = (link, label, includeSeperator = false) => {
+    let seperator;
+    if (includeSeperator) {
+      seperator = (
```
Spelling: `separator`

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.js:
```diff
@@ -0,0 +1,188 @@
+      minWidth: `${minWidth}em`
+    };
+    return (
+      <React.Fragment key={link}>
```
Not sure how this happened, but I got this error due to conflicting keys:

![image](https://user-images.githubusercontent.com/1238659/36004900-d9ff35ca-0ce9-11e8-9823-c06324f15ffa.png)

nreese replied:
Good catch, looks like if you change the name of a saved object than the links are the same but the labels are different. I will make the react key be a combo of both to ensure no duplicate keys.

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.js:
```diff
@@ -0,0 +1,188 @@
+    const openPopoverComponent = (
+      <EuiLink onClick={this.onButtonClick.bind(this)}>
+        <EuiTextColor color="subdued">
+          {`+${dropdownLinks.length}`}
```
Personally, I feel like this would be better spelled out, e.g. "5 more links".  The +{number} isn't quite intuitive to me. @gchaps Thoughts?

![image](https://user-images.githubusercontent.com/1238659/36004934-0e7bb102-0cea-11e8-8b3b-ef0f0ff1d972.png)

gchaps replied:
I agree, @cjcenizal "5 more links" is a better way to convey the meaning.

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.js:
```diff
@@ -0,0 +1,188 @@
+              Recently accessed
+            </EuiTextColor>
+          </p>
+        </EuiText>
```
This looks a little tight. How about a small spacer beneath this title?

![image](https://user-images.githubusercontent.com/1238659/36004969-29f24a22-0cea-11e8-8004-a655511ee8ca.png)

nreese replied:
good idea. Spacer has been added.

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.test.js:
```diff
@@ -0,0 +1,35 @@
+};
+
+test('should render dropdown when recently accessed list size exceeds NUM_LONG_LINKS', () => {
+  const component = shallow(<RecentlyAccessed
```
The snapshots for these tests are massive. How do you feel about using `findTestSubject` to find the specific elements you're looking for instead of snapshots?

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.less:
```diff
@@ -0,0 +1,9 @@
+  text-overflow: ellipsis;
+}
+
+.dropwdownLink {
```
Because these names are global, can we scope them with a prefix to avoid a naming collision? E.g. `recentlyAccessedLongLink` and `recentlyAccessedDropdownLink`.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15657/)

formgeist commented:
Looking great, @nreese! Found some weirdness where the text is cut off, so g's, y's and j's are not look displaying in full. As @cjcenizal found already, I'd just remove the dots when the layout moves to vertical list in the breakpoint, but not a priority.

![screenshot 2018-02-09 11 14 35](https://user-images.githubusercontent.com/4104278/36022832-b6d77048-0d8a-11e8-9362-a67c5a5a0b00.png)

nreese commented on src/core_plugins/kibana/public/home/components/recently_accessed.less:
```diff
@@ -0,0 +1,26 @@
+@media only screen and (max-width: 768px) {
```
@cjcenizal Is there a variable that I can use from EUI?

formgeist commented:
Looking good! I have two things; it seems the "more" popover placement isn't shifted when the responsive breakpoints puts everything on th left.

![screenshot 2018-02-09 14 32 29](https://user-images.githubusercontent.com/4104278/36030108-2df67cf8-0da6-11e8-8796-7a5a326a38d1.png)

Also, could we change the grey dot colour to use `#d9d9d9` (`$globalColorLightGray`), so they're a little more subdued. Thanks!

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15666/)

alexfrancoeur commented:
@nreese looking good!  A couple of quick comments.

Is it at all possible to always right align the context menu if it's present? Or are we specifically presenting this way because of dynamic screen widths?

<img width="1668" alt="screen shot 2018-02-12 at 10 12 07 am" src="https://user-images.githubusercontent.com/25181823/36103726-2922ef74-0fde-11e8-895a-a586f459babb.png">

From an accessibility standpoint, I seem to get stuck in the context menu when tabbing through.
![feb-12-2018 10-16-48](https://user-images.githubusercontent.com/25181823/36103968-c3643354-0fde-11e8-9a96-3ced958f1274.gif)

I'm not sure if this is related to adding support for Timelion sheets, but I'm running into this error when attempting to save a new (and empty) sheet.
<img width="1673" alt="screen shot 2018-02-12 at 10 20 12 am" src="https://user-images.githubusercontent.com/25181823/36104031-e20744e0-0fde-11e8-9fa5-d1274fc79313.png">

alexfrancoeur commented:
Also @formgeist I like the new design. In the previous one, we had a button on the top right hand corner for index pattern creation. This was key in helping users navigate to index pattern creation. Do we have any concerns around pushing this link to the bottom and making it less prevalent?

cjcenizal commented:
@alexfrancoeur FYI popovers deliberately trap focus as an accessibility feature. You have to hit ESC to close it. This way as you can tab without worrying about whether you're in still in the popover or somehow outside of it.

alexfrancoeur commented:
ah, wonderful. Thanks @cjcenizal for the quick response 😄 I should have known you and the team already made sure this component was accessible 👍

formgeist commented:
@alexfrancoeur My reasoning for moving it below the Add data options was that I felt it was getting very noisey in that top with the title, subtitle etc. Is there a case where the Add data options wouldn't show up? I can't remember the different configurations.

alexfrancoeur commented:
@formgeist nope, we should at a minimum, always have three cards. I'm fine with moving it, just wanted to better understand why. Thanks!

formgeist commented:
@alexfrancoeur OK, that's good - IMO it stands out a little more than before, even if it sits a little lower on the page.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15742/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15745/)

kobelb commented on src/core_plugins/kibana/public/home/components/add_data.js:
```diff
@@ -0,0 +1,188 @@
+
+    let apmCard;
+    if (!isCloudEnabled) {
+      apmCard = (
```
I know we're only refactored this code to be here, but I just realized that APM is hard-coded here, even though it's not a part of OSS Kibana, wasn't there a pluggable/registry-oriented way we can accomplish this?

nreese replied:
There is no plug-able way to add cards to the home page.

kobelb replied:
Doesn't that mean that APM always shows up even if we don't have x-pack installed?

Stacey-Gammon replied:
Yep.  APM has an open source side, and you can view the data collected using OSS Kibana dashboards.  The home page tutorial walks you through that, even though there is no APM app inside OSS Kibana (only the x-pack version).  @nreese is working on a PR that will link to either the OSS dashboard, or the APM app, from the tutorial depending on whether x-pack is installed or not (https://github.com/elastic/kibana/pull/16818)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15756/)

nreese commented:
@cjcenizal I have updated the history logic to remove duplicates by object id so renaming an object will no longer put duplicates in the list. I also updated the recently_accessed tests to avoid creating huge snapshots.

You will have to remove your old local storage variable `kibana.history.recentlyAccessed` since the value changed a little.

Let me know if anything else needs changing

gchaps commented:
@nreese How about changing "Recently accessed" to "Recently viewed"?  It's friendlier and better conveys the action.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15765/)

alexfrancoeur commented:
Looks great! (Unique visitors is not a duplicate, ones a dashboard and ones visualize)

<img width="1233" alt="screen shot 2018-02-13 at 2 06 49 pm" src="https://user-images.githubusercontent.com/25181823/36168458-2e5421e0-10c7-11e8-914f-e8412051238a.png">

Open question to the group. I clearly have space for more than 5 links in that screenshot above. Should we limit to 5 when there's white space?

gchaps commented:
"4 more links" doesn't sound quite right to me.  Does "4 more views" work?

nreese commented:
How about just "4 more"?

gchaps commented:
Yup, that works too.  It's in a panel with the other recently-viewed items, so users should make the connection.

cjcenizal commented on src/core_plugins/kibana/public/home/components/home.js:
```diff
@@ -49,173 +39,31 @@ export function Home({ addBasePath, directories, isCloudEnabled }) {
+  let recentlyAccessedPanel;
+  if (recentlyAccessed.length > 0) {
+    recentlyAccessedPanel = (
+      <React.Fragment>
```
In general, I think we're importing submodules instead of referring to them as properties on the main module. I think this is more "ES6-y"?

```jsx
import React, { Fragment } from 'react';

<Fragment />
```

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.js:
```diff
@@ -0,0 +1,210 @@
+
+export const NUM_LONG_LINKS = 5;
+
+export class RecentlyAccessed extends React.Component {
```
Same idea here, `import React, { Component } from 'react';`

cjcenizal commented on src/core_plugins/kibana/public/home/components/recently_accessed.test.js:
```diff
@@ -0,0 +1,77 @@
+    />);
+
+    const moreRecentlyAccessed = findTestSubject(component, 'openMoreRecentlyAccessedPopover');
+    expect(0).toBe(moreRecentlyAccessed.length);
```
I think these assertions are reversed. I'd expect us to assert that the length of `moreRecentlyAccessed` would be 0 rather than the other way around.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/72/)

nreese commented:
I changed the label from `4 more links` to `4 more`. I also increased the history size from 10 to 20. 

Are there any other objections/concerns for this PR. I am hoping to merge before the end of the week.

alexfrancoeur commented:
LGTM

On Tue, Feb 20, 2018, 5:59 PM Nathan Reese <notifications@github.com> wrote:

> I changed the label from 4 more links to 4 more. I also increased the
> history size from 10 to 20.
>
> Are there any other objections/concerns for this PR. I am hoping to merge
> before the end of the week.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/kibana/pull/16556#issuecomment-367151122>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AYA-f5Lk8Ya030aV3vcz6mdKBHHFkKZ5ks5tW05ZgaJpZM4R7qG5>
> .
>

formgeist commented:
@nreese I'm having issues with long saved query names in the recently view panel. Is there a max-width for the items?

![screenshot 2018-02-21 11 24 37](https://user-images.githubusercontent.com/4104278/36475078-e01b2576-16f9-11e8-83c3-d5c621878acc.png)

nreese commented:
@formgeist I can reproduce the problem you are seeing. Their is a max-width of 300px but may be that is too big?

```
.recentlyAccessedItem {
  overflow: hidden;
  max-width: 300px;
}
```

cchaos commented:
You might want to revisit this pen: https://codepen.io/clhorn99/pen/KQaMQj

nreese commented:
@cchaos @formgeist Looks like adding a max-width to the flex item containing the links solved the problem.

Are there any other concerns/issues that need to be addressed?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/152/)

alexfrancoeur commented:
:shipit:

Stacey-Gammon commented on src/core_plugins/kibana/public/home/components/recently_accessed.js:
```diff
@@ -0,0 +1,206 @@
+    };
+  }
+
+  onButtonClick() {
```
can avoid the .bind(this) for this function and closePopover by using the same notation as renderDropdown, e.g.:

```
closePopover = () => {
```

Stacey-Gammon commented on src/core_plugins/kibana/public/home/components/recently_accessed.test.js:
```diff
@@ -0,0 +1,77 @@
+      moreRecentlyAccessed.simulate('click');
+
+      let i = 0;
+      while(i < numberOfRecentlyAccessed) {
```
need a space after while (I thought the linter would catch that?)

nreese commented:
@cjcenizal @chrisronline Had to bump `enzyme-to-json` to `3.3.0` for [React.Fragment enzyme support](https://github.com/adriantoine/enzyme-to-json/issues/89)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/353/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/364/)

nreese commented:
@Stacey-Gammon Do you want to take another look at this before merging?

Stacey-Gammon commented:
Looks good!
